enclave-runner/src/usercalls/mod.rs

```diff
@@ -76,148 +83,307 @@ impl<T> Write for ReadOnly<T> {
+    fn new<S: AsyncStream>(s: Box<S>) -> Self {
+        AsyncStreamContainer {
+            inner: Lock::new(AsyncStreamAdapter {
+                stream: s,
+                read_queue: VecDeque::new(),
+                write_queue: VecDeque::new(),
+                flush_queue: VecDeque::new(),
+            }),
+        }
     }
 
-    fn write(&self, buf: &[u8]) -> IoResult<usize> {
-        Write::write(&mut { self }, buf)
+    async fn async_read(&self, buf: &mut [u8]) -> IoResult<usize> {
+        tokio::prelude::future::poll_fn(|| match self.inner.clone().poll_lock() {
+            Async::NotReady => Ok(Async::NotReady),
+            Async::Ready(mut adapter) => adapter.poll_read(buf),
+        })
+        .compat()
+        .await
     }
 
-    fn flush(&self) -> IoResult<()> {
-        Write::flush(&mut { self })
+    async fn async_read_alloc(&self) -> IoResult<Vec<u8>> {
+        let mut vec: Vec<u8> = vec![0; 8192];
+
+        let res = tokio::prelude::future::poll_fn(|| match self.inner.clone().poll_lock() {
+            Async::NotReady => Ok(Async::NotReady),
+            Async::Ready(mut adapter) => adapter.poll_read(vec.as_mut_slice()),
+        })
+        .compat()
+        .await;
+
+        match res {
+            Ok(size) => {
+                vec.resize(size, 0);
+                Ok(vec)
+            }
+            Err(e) => Err(e),
+        }
     }
-}
 
-/// This trait is mostly same as `std::io::Read` + `std::io::Write` except that it takes an immutable reference to the source.
-pub trait SyncStream: 'static + Send + Sync {
-    /// Read some data from stream, letting the callee choose the amount.
-    fn read_alloc(&self) -> IoResult<Vec<u8>> {
-        let mut buf = vec![0; 8192];
-        let len = self.read(&mut buf)?;
-        buf.resize(len, 0);
-        Ok(buf)
+    async fn async_write(&self, buf: Vec<u8>) -> IoResult<usize> {
+        tokio::prelude::future::poll_fn(|| match self.inner.clone().poll_lock() {
+            Async::NotReady => Ok(Async::NotReady),
+            Async::Ready(mut adapter) => adapter.poll_write(buf.as_slice()),
+        })
+        .compat()
+        .await
     }
 
-    /// Same as `std::io::Read::read`, except that it takes an immutable reference to the source.
-    fn read(&self, buf: &mut [u8]) -> IoResult<usize>;
-    /// Same as `std::io::Write::write` , except that it takes an immutable reference to the source.
-    fn write(&self, buf: &[u8]) -> IoResult<usize>;
-    /// Same as `std::io::Write::flush` , except that it takes an immutable reference to the source.
-    fn flush(&self) -> IoResult<()>;
+    async fn async_flush(&self) -> IoResult<()> {
+        tokio::prelude::future::poll_fn(|| match self.inner.clone().poll_lock() {
+            Async::NotReady => Ok(Async::NotReady),
+            Async::Ready(mut adapter) => adapter.poll_flush(),
+        })
+        .compat()
+        .await
+    }
 }
 
+struct AsyncListenerContainer {
+    inner: Mutex<Box<dyn AsyncListener>>,
```

Comment:
Is there a reason this is a `futures::Mutex` instead of a `tokio::Lock`? I'd prefer to only use one of them throughout the file.

Reply:
futures::Mutex was the thing that least amount of changes, hence I went for that. But for some things tokio::Lock was better (eg our stdin implementation). 
I've changed all the mutex to Lock and used a macro to unwrap.